labs/b5.md

```diff
@@ -0,0 +1,371 @@
+
+But it's much easier for us to remember names like google, facebook, or coolmath-games.com. So out of this conflict the Domain Name System (DNS) was born as a compromise between machine friendly IP addresses and human friendly domain names.
+
+DNS is a system that maps a domain name like google.com to `172.217.6.78`. When you query for google.com your computer sends out a DNS query for google.com to a DNS server. Assuming things are properly configured and google.com has a valid corresponding address you will receive a response from an authoritative server that essentially says "google.com has IP address `x.x.x.x`".
+
+Now let's flush out this black magic a bit...
+
+## DNS Resolvers
+A DNS resolver on the client is responsible for performing DNS queries. A resolver refers to `/etc/resolv.conf` for nameserver information. The resolver then issues a query to the nameserver for the domain name. Additionally, `/etc/hosts` stores static entries for resolving hostnames allowing a sysadmin to specify certain mappings manually. 
+
+To resolve a query the nameserver breaks down the domain name from right to left and issues queries that grow in specificity. Firstly, we note that there are a set of well-known nameservers known as root nameservers at the top of the DNS heirarchy that serve as the first point of contact in resolving a DNS query.  Let's take inst.eecs.berkeley.edu for example. Assuming none of our nameservers have cached data our nameserver will query the root server to find the nameserver for the corresponding Top Level Domain (TLD), which is .edu in this case. The Top Level Domain points to another nameserver which would be authoritative over the next subdomain i.e. berkeley.  This continue on until we've flushed out the entire domain name inst.eecs.berkeley.edu and receive its corresponding IP address.  Of course this isn't the whole story...
+
+There are two major methods that the nameserver may use to resolve a query. 
+
+First is the Iterative DNS Query:
+![alt text][iDNS]
+
+[iDNS]: https://i.imgur.com/qmI1krB.png "Iterative DNS"
+Here the nameserver must issue multiple queries per response from a nameserver until they find the IP address for the given domain name.
+
+Second is the Recursive DNS Query:
+![alt text][rDNS]
+
+[rDNS]: https://i.imgur.com/wP6xaXt.png "Recursive DNS"
+In recursive DNS a query is issued from the nameserver to the root level server and then the root server issues a query to the corresponding TLD server.  This process repeats with each nameserver querying the next one in the chain until we arrive at the lowest level nameserver with the corresponding IP address and the query rolls back.
+
+Notice how DNS is a hierarchy of authoritative servers. This architecture is why DNS managed to scale alongside the internet and decentralizes the administration of DNS.
```

Comment:
Iterative and recursive queries can coexist too, in that you can make a recursive query and then the next resolver answers it iteratively, but I don't know if you want to go into that much detail. Might be cool to mention during the lecture or something? That's commonly done with local recursive DNS servers though if the upstream DNS server is not enabled to be recursive.